src/Features/Core/Portable/Completion/CompletionService.cs

```diff
@@ -206,44 +206,111 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
                     {
-                        // This item is strictly better than the best items we've found so far.
-                        bestItems.Clear();
-                        bestItems.Add(pair);
+                        highestMatchPriorityInBest = pair.item.Rules.MatchPriority;
+                        // No need to remove items with lower MatchPriority from similarItemsWithHigerMatchPriority list,
+                        // we will only add ones with higher value at the end.
                     }
-                    else if (comparison == 0)
+                }
+                else if (comparison < 0)
+                {
+                    // This item is strictly better than the best items we've found so far.
+                    // However, if it's only better in terms of case-sensitivity, we'd like 
+                    // to save the prior best items and consider their MatchPriority later.
+                    similarItemsWithHigherMatchPriority.Clear();
+
+                    if (filterTextContainsNoUpperLetters &&
+                        onlyDifferInCaseSensitivity &&
+                        highestMatchPriorityInBest > pair.item.Rules.MatchPriority) // don't add if this item has higher MatchPriority than all prior best items
                     {
-                        // This item is as good as the items we've been collecting.  We'll return 
-                        // it and let the controller decide what to do.  (For example, it will
-                        // pick the one that has the best MRU index).
-                        bestItems.Add(pair);
+                        similarItemsWithHigherMatchPriority.AddRange(bestItems);
                     }
+
+                    bestItems.Clear();
+                    bestItems.Add(pair);
+                    highestMatchPriorityInBest = pair.item.Rules.MatchPriority;
+                }
+                else
+                {
                     // otherwise, this item is strictly worse than the ones we've been collecting.
-                    // We can just ignore it.
+                    // However, if it's only worse in terms of case-sensitivity, we'd like 
+                    // to save it and consider its MatchPriority later.
+                    if (filterTextContainsNoUpperLetters &&
+                        onlyDifferInCaseSensitivity &&
+                        pair.item.Rules.MatchPriority > highestMatchPriorityInBest)  // don't add if this item doesn't have higher MatchPriority
+                    {
+                        similarItemsWithHigherMatchPriority.Add(pair);
+                    }
+                }
+            }
+
+            // Include those similar items (only worse in terms of case-sensitivity) that have better MatchPriority.
+            foreach (var pair in similarItemsWithHigherMatchPriority)
+            {
+                if (pair.item.Rules.MatchPriority > highestMatchPriorityInBest)
```

Comment:
is this check needed.  i'm a little confsued by it.  by definition, don't similarItemsWithHigherMatchPriority all have a higher match priority?

Reply:
We might add more items to `bestItems` afterwards, which can have higher MatchPriority than earlier ones (i.e. when comparison == 0). However, we don't try to cleanup `similarItemsWithHigherMatchPriority` at that time.

Reply:
so here is the confusing part.  the list already says "withhighermatchpriority", but then you're checking the priority again.  so i think teh list is not named properly.  i think perhaps it should be similarItemsWithWorseCaseSensitivity (or somethign like that).